WalletWasabi/WabiSabi/Client/GreedyDecomposer.cs

```diff
@@ -0,0 +1,43 @@
+using System.Linq;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System;
+using NBitcoin;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class GreedyDecomposer
+	{
+		public GreedyDecomposer(IEnumerable<Money> denominations)
+		{
+			Denominations = denominations.Any()
+				? denominations.OrderByDescending(x => x).ToImmutableList()
+				: throw new ArgumentException($"Argument {nameof(denominations)} has no elements");
+		}
+
+		public ImmutableList<Money> Denominations { get; }
+
+		public IEnumerable<Money> Decompose(Money amount)
+		{
+			var i = 0;
+			var denomination = Denominations[i];
+			while (amount > Money.Zero)
+			{
+				if (denomination <= amount)
+				{
+					yield return denomination;
+					amount -= denomination;
+				}
+				else if (++i < Denominations.Count)
+				{
+					denomination = Denominations[i];
+				}
+				else
+				{
+					yield return amount;
+				}
```

Comment:
```suggestion
```

This return can result in non-standard dust outputs, which would be result in an error from the multiparty tx state stuff, or leak information about precise remaining amounts that can constrain things, since `amount` is not a standard denomination.

if `amount` is larger than some dust threshold, then we can consider introducing some concept of failure to decomposition, but the approach I took for that is that a `Decomposition` is a representation of partial solution, not necessarily a complete one. The natural way to extend one partial solution with another strategy is to calculate the remaining amount, decompose that, and then concatenate/union the two partial solutions.

These edge cases can arise when the user's balance stands out compared to other users (e.g. all other users have 1 input greater than this user's combined inputs, for which there is no real solution, but the opposite situation in which a larger user stands out still leaves that larger user with the option of restricting their used denominations only to those that could be conceivably created by the inputs of the smaller users.